src/elf/parser.rs

```diff
@@ -486,6 +485,59 @@ impl ElfParser {
         }
     }
 
+    /// Perform an operation on each symbol.
+    pub(crate) fn for_each_sym<F, R>(&self, opts: &FindAddrOpts, mut r: R, mut f: F) -> Result<R>
+    where
+        F: FnMut(R, &SymInfo<'_>) -> R,
+    {
+        if let SymType::Variable = opts.sym_type {
+            return Err(Error::with_unsupported("Not implemented"))
+        }
+
+        let mut cache = self.cache.borrow_mut();
+        let shdrs = cache.ensure_shdrs()?;
+        let () = cache.ensure_symtab()?;
+        let () = cache.ensure_str2symtab()?;
+        drop(cache);
+
+        let mut i = 0;
+        loop {
+            let cache = self.cache.borrow_mut();
+            // SANITY: The above `ensure_symtab` ensures we have `symtab`
+            //         available.
+            let symtab = cache.symtab.as_ref().unwrap();
```

Comment:
what about DYNSYM vs SYMTAB ELF sections? For lots of cases there are both... For libbpf we chose to pass it as an argument from outside (see https://github.com/libbpf/libbpf/blob/master/src/elf.c#L308-L312), for example, because it's hard to cleanly "merge" transparently, so it's better to let user do these decisions.

Reply:
Right now the semantics are https://github.com/libbpf/blazesym/blob/424ac9094a4ae81aeb0149797d3e87731f903cf1/src/elf/parser.rs#L271-L279

Which is sub-par at best, but nothing that this change really concerns itself with (and it's on my radar to fix it). Can you elaborate on why the merge is hard, though, please?

Reply:
> Right now the semantics are

This came up before here: https://github.com/libbpf/blazesym/pull/190#discussion_r1212125734. So it hasn't been forgotten.

Reply:
For example, in libbpf's case, for DYNSYM symbols we additionally fetch GNU's verneed (or whatever it's called, don't remember) extra information. So there are differences that only user might know. It felt better for libbpf case to let user drive which symbol table to iterate, instead of picking one or the other automatically.

Reply:
We don't currently support reporting of any such information. We don't do it in the existing `lookup` API nor do we do it here. Yes, we may make more of the reporting configurable down the line, but I don't think this should all be front loaded for no reason and without any user in sight.
Right now this API uses the same data set as the `lookup` API, which is entirely reasonable from my point of view and both APIs will benefit from more sophisticated logic of symbol discovery. But again, I don't see why that has to be done here and now. What do you think?